cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,279 @@
+	fc := &FirewallController{
+		kubeClient: kubeClient,
+		client:     client,
+		fwManager:  fwManager,
+		fwCache:    firewallCache{},
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc, nil
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		fc.fwCache.mu.RLock()
+		targetFirewall := fc.fwCache.firewall
+		fc.fwCache.mu.RUnlock()
+		currentFirewall, err := fc.fwManager.Get(ctx)
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if currentFirewall != nil {
+			if cmp.Equal(currentFirewall, targetFirewall) {
+				return
+			}
+		}
+		err = fc.ensureReconciledFirewall(ctx)
+		if err != nil {
+			klog.Errorf("failed to reconcile worker firewall: %s", err)
+		}
+
+	}, runFrequency, stopCh)
+}
+
+// Get returns the current CCM worker firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if fm.fwCache.firewall != nil {
+		fm.fwCache.mu.RLock()
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fm.fwCache.firewall.ID)
+		fm.fwCache.mu.RUnlock()
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("failed to retrieve firewall by ID: %s", err)
+		}
+		if resp != nil && resp.StatusCode == 404 {
+			klog.Infof("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return cmp.Equal(fw.Name, fm.workerFirewallName)
+	}
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %s", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the CCM worker firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	targetFirewall := fm.fwCache.firewall
+	fm.fwCache.mu.RUnlock()
+
+	fwSvc := fm.client.Firewalls
+
+	fm.fwCache.mu.RLock()
+	isEqual := cmp.Equal(targetFirewall.InboundRules, svcInboundRules)
+	fm.fwCache.mu.RUnlock()
+
+	if !isEqual {
+		currentFirewall, err := fm.Get(ctx)
```

Comment:
So, use the `targetFirewall.ID` as the argument to update. And if it isn't found, create it. Because the target firewall is what we want.

Reply:
Dang, I just realized I missed one use case: we might not have a user ID if this is the first run after CCM freshly started up (and a firewall may possibly not even exist yet). So just in that case, we should invoke `Get()` in `Set()` and determine the next step based on the result: if we do get a firewall (and along with that a corresponding ID), we know that we can run an _Update_ operation; otherwise, it must be _Create_.

Sorry for the confusion -- the state handling really makes things complicated (as expected).